typescript-ignorant/TypeScriptParser.g4

```diff
@@ -0,0 +1,838 @@
+ * Software is furnished to do so, subject to the following
+ * conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+parser grammar TypeScriptParser;
+
+options {
+    tokenVocab=TypeScriptLexer;
+    superClass=TypeScriptBaseParser;
+}
+
+// SupportSyntax
+
+initializer
+    : '=' singleExpression
+    ;
+
+bindingPattern
+    : (arrayLiteral | objectLiteral)
+    ;
+
+// TypeScript SPart
+// A.1 Types
+
+typeParameters
+    : '<' typeParameterList? '>'
+    ;
+
+typeParameterList
+    : typeParameter (',' typeParameter)*
+    ;
+
+typeParameter
+    : Identifier constraint?
+    | typeParameters
+    ;
+
+constraint
+    : Extends type_
+    ;
+
+typeArguments
+    : '<' typeArgumentList? '>'
+    ;
+
+typeArgumentList
+    : typeArgument (',' typeArgument)*
+    ;
+
+typeArgument
+    : type_
+    ;
+
+type_
+    : unionOrIntersectionOrPrimaryType
+    | conditionalType
+    | functionType
+    | constructorType
+    | typeGeneric
+    | StringLiteral
+    ;
+
+conditionalType
+    : primaryType Extends primaryType '?' type_ ':' type_
+    ;
+
+unionOrIntersectionOrPrimaryType
+    : unionOrIntersectionOrPrimaryType '|' unionOrIntersectionOrPrimaryType # Union
+    | unionOrIntersectionOrPrimaryType '&' unionOrIntersectionOrPrimaryType # Intersection
+    | primaryType # Primary
+    ;
+
+primaryType
+    : '(' type_ ')'                                 # ParenthesizedPrimType
+    | predefinedType                                # PredefinedPrimType
+    | typeReference                                 # ReferencePrimType
+    | objectType                                    # ObjectPrimType
+    | primaryType {notLineTerminator()}? '[' ']'    # ArrayPrimType
+    | '[' tupleElementTypes ']'                     # TuplePrimType
+    | typeQuery                                     # QueryPrimType
+    | This                                          # ThisPrimType
+    | typeReference Is primaryType                  # RedefinitionOfType
+    | Keyof primaryType                             # KeyofType
+    ;
+
+predefinedType
+    : Any
+    | Number
+    | Boolean
+    | String
+    | Symbol
+    | Void
+    | Never
+    ;
+
+typeReference
+    : typeName ( typeIncludeGeneric | typeGeneric | '[' Identifier ']')?
+    ;
+
+// I tried recursive include, but it's not working.
+// typeGeneric
+//    : '<' typeArgumentList typeGeneric?'>'
```

Comment:
Why is it not working, what error?

Have you tried that `'<' (typeArgumentList | typeArgumentList typeGeneric) '>'`?

Reply:
I don't know! @loonydev wrote it. :smiley:

Reply:
Hi, 
I don't remember all info, but you can check in my PR for this.
Of course it's better to use recursive, but if I use recursive it ignore on < typeArgumentList < typeArgumentList (>>) - error here. 

That's why I inserted this kind of "fix"

Reply:
> but if I use recursive it ignore on < typeArgumentList < typeArgumentList (>>) - error here

That's why I suggested removing `'>>'` token below.